### PR TITLE
Fix nullable int bug

### LINKED ISSUE #1538: Outlier Samples Detection & Drift can't handle Pandas Int types
Came from external source - we have no access to the original data.
However, I do know they used "Int8" dtype which has Nones in it, and that might be the root cause (but not for sure).

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-41-253b0694d00d> in <module>
      1 from deepchecks.tabular.checks import OutlierSampleDetection
      2 check = OutlierSampleDetection()
----> 3 check.run(ds_train)

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/deepchecks/tabular/base_checks.py in run(self, dataset, model, **kwargs)
     56             dataset,
     57             model=model,
---> 58             **kwargs
     59         ))
     60 

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/deepchecks/tabular/base_checks.py in wrapped(*args, **kwargs)
     34     @wraps(func)
     35     def wrapped(*args, **kwargs):
---> 36         result = func(*args, **kwargs)
     37         return check_instance.finalize_check_result(result)
     38 

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/deepchecks/tabular/checks/data_integrity/outlier_sample_detection.py in run_logic(self, context, dataset_type)
    106             cat_data=df[dataset.cat_features].iloc[:DATASET_TIME_EVALUATION_SIZE],
    107             numeric_data=df[dataset.numerical_features].iloc[:DATASET_TIME_EVALUATION_SIZE],
--> 108             num_neighbours=int(min(np.sqrt(DATASET_TIME_EVALUATION_SIZE), num_neighbours)))
    109         predicted_time_to_run_in_seconds = ((time.time() - start_time) / 130000) * (df.shape[0] ** 2)
    110         if predicted_time_to_run_in_seconds > self.timeout > 0:

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/deepchecks/utils/gower_distance.py in calculate_nearest_neighbours_distances(cat_data, numeric_data, num_neighbours)
     89     # handle categorical - transform to an ordinal numpy array
     90     enc = OrdinalEncoder()
---> 91     cat_data = enc.fit_transform(cat_data.fillna(value=''.join(random.choices(string.printable, k=16))))
     92     # handle numerical - calculate ranges per feature and fill numerical nan to minus np.inf
     93     numeric_data = np.asarray(numeric_data.fillna(value=np.nan))

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/util/_decorators.py in wrapper(*args, **kwargs)
    309                     stacklevel=stacklevel,
    310                 )
--> 311             return func(*args, **kwargs)
    312 
    313         return wrapper

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/frame.py in fillna(self, value, method, axis, inplace, limit, downcast)
   5180             inplace=inplace,
   5181             limit=limit,
-> 5182             downcast=downcast,
   5183         )
   5184 

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/generic.py in fillna(self, value, method, axis, inplace, limit, downcast)
   6381             elif not is_list_like(value):
   6382                 new_data = self._mgr.fillna(
-> 6383                     value=value, limit=limit, inplace=inplace, downcast=downcast
   6384                 )
   6385             elif isinstance(value, ABCDataFrame) and self.ndim == 2:

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/internals/managers.py in fillna(self, value, limit, inplace, downcast)
    409     def fillna(self: T, value, limit, inplace: bool, downcast) -> T:
    410         return self.apply(
--> 411             "fillna", value=value, limit=limit, inplace=inplace, downcast=downcast
    412         )
    413 

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/internals/managers.py in apply(self, f, align_keys, ignore_failures, **kwargs)
    325                     applied = b.apply(f, **kwargs)
    326                 else:
--> 327                     applied = getattr(b, f)(**kwargs)
    328             except (TypeError, NotImplementedError):
    329                 if not ignore_failures:

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/internals/blocks.py in fillna(self, value, limit, inplace, downcast)
   1570         self, value, limit=None, inplace: bool = False, downcast=None
   1571     ) -> list[Block]:
-> 1572         values = self.values.fillna(value=value, limit=limit)
   1573         return [self.make_block_same_class(values=values)]
   1574 

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/arrays/masked.py in fillna(self, value, method, limit)
    174                 # fill with value
    175                 new_values = self.copy()
--> 176                 new_values[mask] = value
    177         else:
    178             new_values = self.copy()

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/arrays/masked.py in __setitem__(self, key, value)
    186         if _is_scalar:
    187             value = [value]
--> 188         value, mask = self._coerce_to_array(value)
    189 
    190         if _is_scalar:

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/arrays/integer.py in _coerce_to_array(self, value)
    332 
    333     def _coerce_to_array(self, value) -> tuple[np.ndarray, np.ndarray]:
--> 334         return coerce_to_array(value, dtype=self.dtype)
    335 
    336     def astype(self, dtype, copy: bool = True) -> ArrayLike:

~/anaconda3/envs/mxnet_latest_p37/lib/python3.7/site-packages/pandas/core/arrays/integer.py in coerce_to_array(values, dtype, mask, copy)
    202 
    203     elif not (is_integer_dtype(values) or is_float_dtype(values)):
--> 204         raise TypeError(f"{values.dtype} cannot be converted to an IntegerDtype")
    205 
    206     if mask is None:

TypeError: <U16 cannot be converted to an IntegerDtype

```

### PR DESCRIPTION
**Pull request by JKL98ISR**
couldn't find a bug in EMD :( closes #1538 if we can't reproduce

### CODE REVIEW

File: deepchecks/utils/gower_distance.py [hunk 1]
```diff
@@ -88,9 +85,11 @@ def calculate_nearest_neighbours_distances(cat_data: pd.DataFrame, numeric_data:
     distances, indexes = np.zeros((num_samples, num_neighbours)), np.zeros((num_samples, num_neighbours))
     # handle categorical - transform to an ordinal numpy array
     enc = OrdinalEncoder()
-    cat_data = enc.fit_transform(cat_data.fillna(value=''.join(random.choices(string.printable, k=16))))
+    if not cat_data.empty:
+        cat_data = cat_data.apply(lambda x: pd.factorize(x)[0])
```
Comment (matanper) [inline deepchecks/utils/gower_distance.py:89]: maybe
```
cat_data, _ = pd.factorize(cat_data)
```

**Review by nirhutnik — APPROVED** [review]

File: tests/conftest.py [hunk 2]
```diff
@@ -115,21 +115,28 @@ def empty_df():
 @pytest.fixture(scope='session')
 def kiss_dataset_and_model():
     """A small and stupid dataset and model to catch edge cases."""
-    def string_to_length(data):
+    def string_to_length(data: pd.DataFrame):
         data = data.copy()
         data['string_feature'] = data['string_feature'].apply(len)
         return data
 
+    def un_nany(data: pd.DataFrame):
```
Comment (TheSolY) [inline tests/conftest.py:123]: the name un_nany is cute but not clear, should be fillna_string_feature

File: tests/conftest.py [hunk 3]
```diff
@@ -115,21 +115,28 @@ def empty_df():
...
             'binary_feature': [0, 1, 1, 0, 0, 1],
             'string_feature': ['ahhh', 'no', 'weeee', 'arg', 'eh', 'E'],
+            'numeric_feature': pd.array([4, np.nan, 7, 3, 2, np.nan], dtype='Int64'),
             'numeric_label': [3, 1, 5, 2, 1, 1],
         })
     train, test = train_test_split(df, test_size=0.33, random_state=42)
     train_ds = Dataset(train, label='numeric_label', cat_features=['binary_feature'])
     test_ds = Dataset(test, label='numeric_label', cat_features=['binary_feature'])
-    clf = Pipeline([('lengthifier', FunctionTransformer(string_to_length)),
+    clf = Pipeline([('un_nany', FunctionTransformer(un_nany)),
```
Comment (TheSolY) [inline tests/conftest.py:138]: Why use Pipeline instead of df.apply?

File: deepchecks/core/check_utils/whole_dataset_drift_utils.py [hunk 4]
```diff
@@ -58,6 +59,16 @@ def run_whole_dataset_drift(train_dataframe: pd.DataFrame, test_dataframe: pd.Da
                                                         random_state=random_state,
                                                         test_size=test_size)
 
+    for df in [x_train, x_test]:
+        df: pd.DataFrame
+        for column in df.columns:
+            df[column] = df[column].astype(float) if is_integer_dtype(df[column]) else df[column]
```
Comment (TheSolY) [inline deepchecks/core/check_utils/whole_dataset_drift_utils.py:65]: Why use loop instead of pandas where? should go something like df.where(is_integer_type(df), df.astype(float), df)

File: tests/conftest.py (hunk 3 above)
Comment (JKL98ISR) [reply tests/conftest.py:138]: so the df will still have those values

**Review by matanper — APPROVED** [review]

File: deepchecks/core/check_utils/whole_dataset_drift_utils.py (hunk 4 above)
Comment (JKL98ISR) [reply deepchecks/core/check_utils/whole_dataset_drift_utils.py]: fixed no he likes it

**Review by ItayGabbay — APPROVED** [review]
